// components/Tree/Tree.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './Tree.css';

export class Node extends Component {
  static propTypes = {
    label: PropTypes.string,
    children: function (props, propName, componentName) {
      React.Children.forEach(props[propName], child => {
        console.log('inside node... isNode: ', child.type === Node);
        if (child.type !== Node) {
          throw new Error('<Node> can only have <Node>\'s as children. ')
        }
      });
    }
  }
  render() {
    return (<div>div</div>);
  }
}

class Tree extends Component {
  static propTypes = {
    root: PropTypes.string,
    children: function (props, propName, componentName) {
      React.Children.forEach(props[propName], child => {
        console.log('inside tree... isNode: ', child.type === Node);
        if (child.type !== Node) {
          throw new Error('<Tree> can only have <Node>\'s as children. ')
        }
      });
    }
  }
  subtree(nodes) {
    let obj = [];
    React.Children.map(nodes, (node) => {
      if(node.type === Node) {
        if(node.props.children) {
          obj.push(
            <li>
              <a>{node.props.label || 'node'}</a>
              {this.subtree(node.props.children)}
            </li>
          );
        } else {
          obj.push(<li><a>{node.props.label || 'node'}</a></li>);
        }
      }
    });
    return (<ul>{obj}</ul>);
  }
  render() {
    return (
      <div className="container">
        <div className="tree">
          <ul>
            <li>
              <a>{'root' || this.props.root}</a>
              {this.subtree(this.props.children)}
            </li>
          </ul>
        </div>
      </div>
    );
  }
}

Tree.Node = Node;

export default Tree;
